libslide/src/bignum/mul.rs

```diff
@@ -0,0 +1,146 @@
+        let is_neg: bool = self.is_neg ^ rhs.is_neg;
+        // total length of return vector is p+q+1
+        let total_length = self.dec.len() + self.int.len() + rhs.dec.len() + rhs.int.len();
+        // need self and rhs dec lengths for split later
+        let dec_len = self.dec.len() + rhs.dec.len();
+        // we need to make the vectors the same length at both the decimal and integer level.
+        let (mut lhs_dec, mut rhs_dec) = match self.dec.len().cmp(&rhs.dec.len()) {
+            Ordering::Greater => {
+                let mut rhs_dec = vec![0; self.dec.len()];
+                for (i, e) in rhs.dec.iter().enumerate() {
+                    rhs_dec[i] = *e
+                }
+                (self.dec, rhs_dec)
+            }
+            Ordering::Less => {
+                let mut lhs_dec = vec![0; rhs.dec.len()];
+                for (i, e) in self.dec.iter().enumerate() {
+                    lhs_dec[i] = *e
+                }
+                (lhs_dec, rhs.dec)
+            }
+            Ordering::Equal => (self.dec, rhs.dec),
+        };
+
+        let (mut lhs_int, mut rhs_int) = match self.int.len().cmp(&rhs.int.len()) {
+            Ordering::Greater => {
+                let mut rhs_int = vec![0; self.int.len()];
+                for (i, e) in rhs.int.iter().enumerate() {
+                    rhs_int[i] = *e
+                }
+                (self.int, rhs_int)
+            }
+            Ordering::Less => {
+                let mut lhs_int = vec![0; rhs.int.len()];
+                for (i, e) in self.int.iter().enumerate() {
+                    lhs_int[i] = *e
+                }
+                (lhs_int, rhs.int)
+            }
+            Ordering::Equal => (self.int, rhs.int),
+        };
+
+        // need to make vectors mod 2 for fft
+        while rhs_int.len() + rhs_dec.len() < total_length {
+            rhs_int.push(0);
+            lhs_int.push(0);
+        }
+        while !ispowerof2((rhs_int.len() + rhs_dec.len()) as f64) {
+            rhs_int.push(0);
+            lhs_int.push(0);
+        }
+        rhs_dec = rhs_dec.iter().rev().cloned().collect();
+        lhs_dec = lhs_dec.iter().rev().cloned().collect();
+        rhs_dec.append(&mut rhs_int);
+        lhs_dec.append(&mut lhs_int);
+        let rhs_len = rhs_dec.len();
+        let rhs_cplx = recast_user_vec(rhs_dec).unwrap();
+        let lhs_cplx = recast_user_vec(lhs_dec).unwrap();
+        let mut rhs_fft = fft(rhs_cplx, rhs_len, false);
+        let lhs_fft = fft(lhs_cplx, rhs_len, false);
+
+        for i in 0..rhs_len {
+            rhs_fft[i] = rhs_fft[i] * lhs_fft[i];
+        }
+
+        let res_round = round_complex_vec(ifft(rhs_fft, rhs_len));
+        let res: Vec<u16> = recast_user_vec(res_round).unwrap();
+        let mut dec_vec: Vec<u8> = convert_poly(res);
+        print!("{:?}", dec_vec);
+        if dec_len > 0 {
+            while dec_vec[0] == 0 {
+                dec_vec.remove(0);
+            }
+        }
```

Comment:
Maybe it's better to use [`skip_while`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip_while) to avoid shifting down every time

Reply:
On the other hand, do we care if there are trailing zeros? If this is about formatting, we can just add more code in the fmt implementation. Doing this is kind of expensive because we have to shift things around or allocate more, and makes no difference in terms of the math right

Reply:
I think i just did a neat trick so we can remove them in the end. We still want to remove the zeros because it does slow our processing time down.

Reply:
why does it slow the processing time down?

Reply:
well because if we have extra zeros just sitting in our vector and then we say add it, it still is going to compute through those zeros.